src/reader/warming.rs

```diff
@@ -0,0 +1,384 @@
+        });
+        let warmers = self.pruned_warmers();
+        // Avoid threads (warming as well as background GC) if there are no warmers
+        if warmers.is_empty() {
+            self.prune_searcher_generations();
+        } else {
+            self.start_gc_thread_maybe(this)?;
+            warming_executor(self.num_warming_threads.min(warmers.len()))?
+                .map(|warmer| warmer.warm(searcher), warmers.into_iter())?;
+        }
+        Ok(())
+    }
+
+    /// Attempt to upgrade the weak Warmer references, pruning those which cannot be upgraded.
+    /// Return the strong references.
+    fn pruned_warmers(&mut self) -> Vec<Arc<dyn Warmer>> {
+        let strong_warmers = self
+            .warmers
+            .iter()
+            .flat_map(|weak_warmer| weak_warmer.upgrade())
+            .collect::<Vec<_>>();
+        self.warmers = strong_warmers.iter().map(Arc::downgrade).collect();
+        strong_warmers
+    }
+
+    /// Prune dropped searcher generations from our state.
+    /// Return count of removed generations.
+    fn prune_searcher_generations(&mut self) -> usize {
+        // This can be implemented neatly using Vec::drain_filter() when that is stable.
+        let mut pruned = 0;
+        let mut i = 0;
+        while i < self.searcher_generations.len() {
+            if !self.searcher_generations[i].is_live() {
+                self.searcher_generations.remove(i);
+                pruned += 1;
+            } else {
+                i += 1;
+            }
+        }
+        pruned
+    }
+
+    /// Segment ID set of live searcher generations.
+    fn live_segment_ids(&self) -> HashSet<SegmentId> {
+        self.searcher_generations
+            .iter()
+            .filter(|gen| gen.is_live())
+            .flat_map(|gen| gen.segment_ids.iter().copied())
+            .collect()
+    }
+
+    /// [Warmer::garbage_collect] active warmers if some searcher generation is observed to have been dropped.
+    fn gc_maybe(&mut self) -> bool {
+        if self.prune_searcher_generations() > 0 {
+            let live_segment_ids = self.live_segment_ids();
+            for warmer in self.pruned_warmers() {
+                warmer.garbage_collect(&live_segment_ids);
+            }
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Start GC thread if one has not already been started.
+    fn start_gc_thread_maybe(&mut self, this: &Arc<Mutex<Self>>) -> crate::Result<bool> {
+        if self.gc_thread.is_some() {
+            Ok(false)
+        } else {
+            let weak_inner = Arc::downgrade(this);
+            let handle = std::thread::Builder::new()
+                .name("tantivy-warm-gc".to_owned())
+                .spawn(|| Self::gc_loop(weak_inner))
+                .map_err(|_| {
+                    TantivyError::SystemError("Failed to spawn warmer GC thread".to_owned())
+                })?;
+            self.gc_thread = Some(handle);
+            Ok(true)
+        }
```

Comment:
```suggestion
        if self.gc_thread.is_some() {
            return Ok(false);
        }
        let weak_inner = Arc::downgrade(this);
        let handle = std::thread::Builder::new()
            .name("tantivy-warm-gc".to_owned())
            .spawn(|| Self::gc_loop(weak_inner))
            .map_err(|_| {
               TantivyError::SystemError("Failed to spawn warmer GC thread".to_owned())
             })?;
        self.gc_thread = Some(handle);
        Ok(true)
```

Reply:
Super minor. The code style in tantivy is quite a mess, but we are trying to move to early returns.